core/DataAccess/LogAggregator.php

```diff
@@ -373,35 +377,118 @@ public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit
                         break;
                     }
                 }
-
             }
 
         }
 
         $query = $segment->getSelectQuery($select, $from, $where, $bind, $orderBy, $groupBy, $limit, $offset);
 
+        if (is_array($query) && array_key_exists('sql', $query)) {
+            $query['sql'] = $this->addOriginHintToQuery($query['sql']);
+        }
+
+        return $query;
+    }
+
+    /**
+     * Generate a query / bind array with segment and origin hints for a custom SQL query, similar to generateQuery()
+     * but for use when a query is too complicated to be assembled from SQL fragments and needs to be manually specified
+     *
+     * @param string     $sql       The custom query string
+     * @param array|null $bindOrder Optionally customise the returned bind parameters array by supplying an array
+     *                              which contains a list of standard bind array keys, any order or repetition is
+     *                              supported. eg. [0,1,2,0,1,2] would repeat the standard bind values twice. If omitted
+     *                              then the bind values will be returned once in order: [0,1,2]
+     *
+     * @return array     Array containing the SQL query and bind array, ['sql' => 'SELECT ...', 'bind' => ['abc', 123]]
+     */
+    public function generateQueryCustom(string $sql, ?array $bindOrder = null): array
+    {
+        // Setup bind array
+        $bind = $this->getGeneralQueryBindParams();
+        if ($bindOrder !== null) {
+            $customBind = [];
+            foreach ($bindOrder as $bindKey) {
+                $customBind[] = $bind[$bindKey];
+            }
+            $bind = $customBind;
+        }
+
+        // Apply segment if required.
+        // Ignores enable_segment_cache = 0
```

Comment:
A problem here might be, that the method to remove the temp table again always checks the config flag and thus won't clean up the table correctly if the flag is disabled.

Reply:
I've added an additional private boolean property `forceCheckForSegmentTmpOnCleanup` to the `LogAggregator`  class to cover this possibility, If a custom query is used which also creates a segment temporary table then this flag is set and the cleanup method will still check for the segment temp table even if the segment cache is disabled.